### DISCUSSION TITLE
AlpineJS fails on leading semicolon inside directive

### DISCUSSION THREAD

**Discussion by natepisarski** [alpinejs/alpine#1384] [7. Bugs (archived)]
## The Bug
This report is about a bug I encountered while working with `x-init`, although I presume that the issue exists in any `x-*` directive where Javascript code is evaluated. The actual error message is this:
> Alpine Error: "SyntaxError: expected expression, got ';'"

The directive that makes this fail is:

```javascript
x-init=";console.debug('hello');"
```
Note the leading `;`. This also happens in this case:

```javascript
x-init=";"
```
## The Use Case
Now, why is it useful to have a leading `;`? This is here because I have code that is dynamically adding statements to these directives. To be safe, I had planned on wrapping each statement in `;` so that if one of the statements omitted their trailing `;`, it wouldn't bring down the whole thing.

I also want to note that Javascript's `eval()` does handle this case properly.

## The (sort of) workaround
Now, there is a workaround but it's very hacky and obtuse. It's to define a one-letter function:

```javascript
export const t = () => {};
```

And use this as the wrapper

```javascript
x-init="t(); console.debug('hi');`
```
This code will work as expected, and provide the safety I'm looking for. But I feel like the leading semicolon (or lone semicolon) should be evaluated properly.

**Comment by SimoTod** [comment]
It's a good point. For your purpose, you don't need to wrap the statements in `;`, you just need the trailing one. You can also put them in an array and then use `.join(`;`)`

**Comment by natepisarski** [comment]
Agreed @SimoTod this does work. Without going into boring technical detail, that solution doesn't apply to my case but that probably covers 90% of them. It's a good workaround

**Comment by SimoTod** [comment]
@natepisarski No problems. The reason why it doesn't work is that Alpine doesn't use eval. x-init expect a value that can. be assigned to a variable internally (see https://github.com/alpinejs/alpine/blob/bbfac59a7f1b0b728837cf44102d175e0fd8a90f/src/utils.js#L94). 
Adding different instructions separated by a semicolon is more of an unexpected feature rather than something it was designed for. 

I guess a possible fix would be to trim all the semicolons from the string, not sure if it has any side effects but you can try to submit a PR.

**Comment by eevleevs** [comment]
Would be useful also when working with Prettier. It adds a leading semicolon for something like

```html
<button @click=";(a = b()) && c()"></button>
```

**Reply by SimoTod** [reply]
Is it something you can disable in prettier? Even without the error in console is fixed, the leading semicolon will break other directives which are expecting a single expression returning a value while the semicolon will create 2 directives.

**Reply by eevleevs** [reply]
It has been reported as an issue on https://github.com/prettier/prettier/issues/9060, not sure if it will be fixed. Currently I need to ignore the specific element.

**Reply by ekwoka** [reply]
That definitely seems more like a bug on prettier's end since it doesn't seem to serve any purpose at all there.

But it looks like changing prettier to use `semi-colons` will not force these in based on the comments there. So that is an option...

**Reply by eevleevs** [reply]
> That definitely seems more like a bug on prettier's end since it doesn't seem to serve any purpose at all there.

I would not call it a bug, their code is valid, even if the semicolon has no purpose.

> But it looks like changing prettier to use `semi-colons` will not force these in based on the comments there. So that is an option...

Not if you hate semicolons 😆

**Reply by ekwoka** [reply]
Well, you can make that PR.

Just pre-process expressions to removing leading semis.

I mean it's a bug because they are adding semis even where it can break things and are not needed.

So instead of satisfying your "hate of semis" it's adding extra semis.
